<a id='top'></a>
<img align="center" src="https://d32r1sh890xpii.cloudfront.net/news/718x300/2020-04-09_6vgl92fyfo.jpg" width="200" />
<p style="text-align: center;">
    <font size='5' type='bold'>
        <b>Выбор региона для бурения</b>
    </font>
</p>

Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Выводы и рассуждения получились содержательными, их было интересно читать.
* Радует, что тебе удалось разобраться с техникой бутстрапа.
* Отлично, что доверительный интервал и риск убытков посчитаны верно.
* Я оставил несколько советов, надеюсь, что они будут учтены в будущих проектах.
* Редко, кто сдает эту работу с первого раза, молодец!
* Проект зачтен. Удачи в будущих проектах!

### Содержание

<a href='#section_0'>Задачи проекта</a>

<a href='#section_1'>1. Загрузка и подготовка данных</a>

<a href='#section_2'>2. Обучение и проверка модели</a>

<a href='#section_3'>3. Подготовка к расчёту прибыли</a>

<a href='#section_4'>4. Расчет прибыли и рисков</a>

* <a href='#section_4_1'>Регион 1</a>
* <a href='#section_4_2'>Регион 2</a>
* <a href='#section_4_3'>Регион 3</a>

<a href='#section_5'>5. Общий вывод</a>

<a href='#section_end'>Чек-лист готовности проекта</a>

<a id='section_0'></a>
## Задачи проекта

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

Условия задачи:

* Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
* При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
* Бюджет на разработку скважин в регионе — 10 млрд рублей.
* При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
* После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.


<a id='section_1'></a>
# 1. Загрузка и подготовка данных

**Описание данных.**

Данные геологоразведки трёх регионов находятся в файлах:

* /datasets/geo_data_0.csv. Скачать датасет
* /datasets/geo_data_1.csv. Скачать датасет
* /datasets/geo_data_2.csv. Скачать датасет
* id — уникальный идентификатор скважины;
* f0, f1, f2 — три признака точек (неважно, что они означают, но сами признаки значимы);
* product — объём запасов в скважине (тыс. баррелей).

In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [2]:
rnd_state = 260686

In [3]:
# посмотрим на данные
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv')


In [4]:
# посмотрим на данные
display(geo_data_0.head(3))
display(geo_data_1.head(3))
display(geo_data_2.head(3))

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647

id         f0        f1        f2     product
0  kBEdx -15.001348 -8.276000 -0.005876    3.179103
1  62mP7  14.272088 -3.475083  0.999183   26.953261
2  vyE1P   6.263187 -5.948386  5.001160  134.766305

id        f0        f1        f2    product
0  fwXo0 -1.146987  0.963328 -0.828965  27.758673
1  WJtFt  0.262778  0.269839 -2.530187  56.069697
2  ovLUW  0.194587  0.289035 -5.586433  62.871910

In [5]:
# посмотрим, нет ли пропусков в данных
display(geo_data_0.info())
display(geo_data_1.info())
display(geo_data_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

In [6]:
# посмотрим, нет ли повторяющихся данных
display(geo_data_0.duplicated().sum())
display(geo_data_1.duplicated().sum())
display(geo_data_2.duplicated().sum())

0

0

0

### Вывод
* Данные представлены в виде дробных значений. Есть как положительные, так и отрицательные значения.
* Пропусков, некорректных наименований признаков и типов данных не обнаружено.
* Дубликаты в датасетах также не представлены.

С данными все хорошо, можно перейти к моделированию.

<a href='#top'>Back to top</a>

<div class="alert alert-block alert-success">
<b>Успех:</b> Загрузка и первичный осмотр данных проведены корректно.
</div>

<a id='section_2'></a>
# 2. Обучение и проверка модели

Обучим и проверим модель для `региона 1`

In [7]:
target = geo_data_0['product']
features = geo_data_0.drop(['id', 'product'], axis=1)
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=rnd_state
    )

model = LinearRegression()
model.fit(features, target)
predicted_valid = model.predict(features_valid)
print(f'Средний запас предсказанного сырья по региону 1: {predicted_valid.mean()}')
print(f'RMSE по региону 1: {mean_squared_error(target_valid, predicted_valid)**(1/2)}')    

Средний запас предсказанного сырья по региону 1: 92.44041623324769
RMSE по региону 1: 37.658531418632855


Обучим и проверим модель для `региона 2`

In [8]:
target = geo_data_1['product']
features = geo_data_1.drop(['id', 'product'], axis=1)
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=rnd_state
    )

model = LinearRegression()
model.fit(features, target)
predicted_valid = model.predict(features_valid)
print(f'Средний запас предсказанного сырья по региону 2: {predicted_valid.mean()}')
print(f'RMSE по региону 2: {mean_squared_error(target_valid, predicted_valid)**(1/2)}')    

Средний запас предсказанного сырья по региону 2: 68.65840435668257
RMSE по региону 2: 0.89332151929554


Обучим и проверим модель для `региона 3`

In [9]:
target = geo_data_2['product']
features = geo_data_2.drop(['id', 'product'], axis=1)
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=rnd_state
    )

model = LinearRegression()
model.fit(features, target)
predicted_valid = model.predict(features_valid)
print(f'Средний запас предсказанного сырья по региону 3: {predicted_valid.mean()}')
print(f'RMSE по региону 3: {mean_squared_error(target_valid, predicted_valid)**(1/2)}')    

Средний запас предсказанного сырья по региону 3: 95.01732644975496
RMSE по региону 3: 40.0275077010672


### Вывод
* Cредние значения объема нефти регионов 1 и 3 примерно одинаковые. Отметим также, что модели имеют высокую погрешность.
* В регионе 2 погрешность модели значительно ниже, но запасы при этом также минимальны.

<a href='#top'>Back to top</a>

<div class="alert alert-block alert-success">
<b>Успех:</b> В целом, модели обучены корректно.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> На протяжении всего проекта происходит работа с 3мя одинаковыми таблицами. В связи с этим можно было использовать функции и/или циклы, чтобы избежать копирование кода.
<br> Зачем его нужно избегать? Потому что очень часто оно приводит к ошибкам (например, забыл поменять одну из переменных). Впоследствии такие ошибки очень трудно найти.
</div>

<a id='section_3'></a>
# 3. Подготовка к расчёту прибыли

In [10]:
# Все ключевые значения для расчётов сохраним в отдельных переменных

wells = 500 # точек исследовано при разведке региона
best_wells = 200 # лучших точек для разработки
budget = 10e9 # бюджет на разработку скважин в регионе
revenue_per_1000bbl = 450000 # доход от продажи единицы продукта (1000 баррелей нефти)
risk_threshold = .025 # порог вероятности убытков для оценки рисков и выбора перспективных регионов

# Рассчитаем достаточный объём сырья для безубыточной разработки новой скважины
production_min = budget / (best_wells * revenue_per_1000bbl)
production_min


111.11111111111111

<div class="alert alert-block alert-info">
<b>Совет: </b> В названиях константных переменных лучше использовать только БОЛЬШИЕ буквы. Это соглашение между программистами.
</div>

### Вывод
* Минимальный объем сырья для безубыточной разработки составляет 111 тыс барелей. Данный объем выше средних значений по каждому региону.

<a href='#top'>Back to top</a>

<div class="alert alert-block alert-success">
<b>Успех:</b> Точка безубыточности найдена корректно, с выводом согласен.
</div>

<a id='section_4'></a>
# 4. Расчёт прибыли и рисков
Далее выберем регион с наибольшей средней прибылью.

<a id='section_4_1'></a>
### Регион 1

In [11]:
features = geo_data_0.drop(['product','id'], axis=1)
target = geo_data_0['product']
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, random_state=12345, test_size=0.25) 
    
model = LinearRegression(n_jobs = -1)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

rmse = np.sqrt(np.mean((target_valid-predicted_valid)**2))
predicted_mean = predicted_valid.mean()
score = model.score(features_train,target_train)
target_mean = geo_data_0['product'].mean()

target_df = pd.Series(target_valid.reset_index(drop=True))
predictions_df = pd.Series(predicted_valid)

In [12]:
#Создадим таблицу, чтобы сравнить результаты

scorelist_1 = pd.DataFrame(
    data=[rmse, score, predicted_mean, target_mean], 
    index=['RMSE','SCORE','PREDICTED MEAN', 'TARGET MEAN'], 
    columns=['region 1']
)
scorelist_1

region 1
RMSE            37.579422
SCORE            0.274239
PREDICTED MEAN  92.592568
TARGET MEAN     92.500000

In [13]:
# Техникой bootstrap посчитаем среднюю прибыль, 95% доверительный интервал и риск убытков для первого региона
# возьмём индекс '0', чтобы перевести данные в pd.Series
target = target_df
probabilities = predictions_df

def revenue(target, probabilities, count):
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:count]
    return 450000 * selected.sum() - budget

revenue_1 = revenue(target, probabilities,200)/10**9

state = np.random.RandomState(12345)
values = []
for i in range(1000):
    target_subsample = target.sample(n=500, replace=True, random_state=state)
    probs_subsample = probabilities[target_subsample.index] 
    values.append(revenue(target_subsample, probs_subsample, 200))
    
values = pd.Series(values)
lower = values.quantile(0.025)
mean = values.mean()

risks = (values < 0).mean() * 100
print("Средняя прибыль:", mean/10**9)
print("2.5%-квантиль:", lower/10**9)
print("Риск: {}".format(risks))

confidence_interval = (values.quantile(0.025), values.quantile(0.975))
print('95% доверительный интервал: ', confidence_interval)

Средняя прибыль: 0.4259385269105923
2.5%-квантиль: -0.10209009483793653
Риск: 6.0
95% доверительный интервал:  (-102090094.83793654, 947976353.358369)


<a href='#top'>Back to top</a>

<div class="alert alert-block alert-success">
<b>Успех:</b> Риск посчитан очень "красиво", молодец!
</div>

<div class="alert alert-block alert-success">
<b>Успех:</b> Радует, что для нахождения доверительного интервала использовался именно метод квантилей.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Не очень понятно, зачем нам отдельно 2.5% квантиль.
</div>

<a id='section_4_2'></a>
### Регион 2

In [14]:
features = geo_data_1.drop(['product','id'], axis=1)
target = geo_data_1['product']
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, random_state=12345, test_size=0.25) 
    
model = LinearRegression(n_jobs = -1)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

rmse = np.sqrt(np.mean((target_valid-predicted_valid)**2))
predicted_mean = predicted_valid.mean()
score = model.score(features_train,target_train)
target_mean = geo_data_1['product'].mean()

target_df = pd.Series(target_valid.reset_index(drop=True))
predictions_df = pd.Series(predicted_valid)

In [15]:
#Создадим таблицу, чтобы сравнить результаты

scorelist_2 = pd.DataFrame(
    data=[rmse, score, predicted_mean, target_mean], 
    index=['RMSE','SCORE','PREDICTED MEAN', 'TARGET MEAN'], 
    columns=['region 2']
)
scorelist_2

region 2
RMSE             0.893099
SCORE            0.999625
PREDICTED MEAN  68.728547
TARGET MEAN     68.825000

In [16]:
# Техникой bootstrap посчитаем среднюю прибыль, 95% доверительный интервал и риск убытков для первого региона
# возьмём индекс '0', чтобы перевести данные в pd.Series
target = target_df
probabilities = predictions_df

def revenue(target, probabilities, count):
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:count]
    return 450000 * selected.sum() - budget

revenue_2 = revenue(target, probabilities,200)/10**9

state = np.random.RandomState(12345)
values = []
for i in range(1000):
    target_subsample = target.sample(n=500, replace=True, random_state=state)
    probs_subsample = probabilities[target_subsample.index] 
    values.append(revenue(target_subsample, probs_subsample, 200))
    
values = pd.Series(values)
lower = values.quantile(0.025)
mean = values.mean()

risks = (values < 0).mean() * 100
print("Средняя прибыль:", mean/10**9)
print("2.5%-квантиль:", lower/10**9)
print("Риск: {}".format(risks))

confidence_interval = (values.quantile(0.025), values.quantile(0.975))
print('95% доверительный интервал: ', confidence_interval)

Средняя прибыль: 0.5152227734432899
2.5%-квантиль: 0.06887322537050178
Риск: 1.0
95% доверительный интервал:  (68873225.37050177, 931547591.2570496)


<a href='#top'>Back to top</a>

<a id='section_4_3'></a>
### Регион 3

In [17]:
features = geo_data_2.drop(['product','id'], axis=1)
target = geo_data_2['product']
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, random_state=12345, test_size=0.25) 
    
model = LinearRegression(n_jobs = -1)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

rmse = np.sqrt(np.mean((target_valid-predicted_valid)**2))
predicted_mean = predicted_valid.mean()
score = model.score(features_train,target_train)
target_mean = geo_data_2['product'].mean()

target_df = pd.Series(target_valid.reset_index(drop=True))
predictions_df = pd.Series(predicted_valid)

In [18]:
#Создадим таблицу, чтобы сравнить результаты

scorelist_3 = pd.DataFrame(
    data=[rmse, score, predicted_mean, target_mean], 
    index=['RMSE','SCORE','PREDICTED MEAN', 'TARGET MEAN'], 
    columns=['region 3']
)
scorelist_3

region 3
RMSE            40.029709
SCORE            0.196614
PREDICTED MEAN  94.965046
TARGET MEAN     95.000000

In [19]:
# Техникой bootstrap посчитаем среднюю прибыль, 95% доверительный интервал и риск убытков для первого региона
# возьмём индекс '0', чтобы перевести данные в pd.Series
target = target_df
probabilities = predictions_df

def revenue(target, probabilities, count):
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:count]
    return 450000 * selected.sum() - budget

revenue_3 = revenue(target, probabilities,200)/10**9

state = np.random.RandomState(12345)
values = []
for i in range(1000):
    target_subsample = target.sample(n=500, replace=True, random_state=state)
    probs_subsample = probabilities[target_subsample.index] 
    values.append(revenue(target_subsample, probs_subsample, 200))
    
values = pd.Series(values)
lower = values.quantile(0.025)
mean = values.mean()

risks = (values < 0).mean() * 100
print("Средняя прибыль:", mean/10**9)
print("2.5%-квантиль:", lower/10**9)
print("Риск: {}".format(risks))

confidence_interval = (values.quantile(0.025), values.quantile(0.975))
print('95% доверительный интервал: ', confidence_interval)

Средняя прибыль: 0.43500836278275556
2.5%-квантиль: -0.128880547329789
Риск: 6.4
95% доверительный интервал:  (-128880547.32978901, 969706954.1802679)


In [20]:
# Сведем результаты расчетов по прибыли регионов

print('Прибыль первого региона : {:.2f} млрд рублей'.format(revenue_1))
print('Прибыль второго региона : {:.2f} млрд рублей'.format(revenue_2))
print('Прибыль третьего региона : {:.2f} млрд рублей'.format(revenue_3))

Прибыль первого региона : 3.32 млрд рублей
Прибыль второго региона : 2.42 млрд рублей
Прибыль третьего региона : 2.71 млрд рублей


In [21]:
# Сведем результаты моделирования по метрикам

scorelist = pd.concat([scorelist_1, scorelist_2, scorelist_3], axis=1, join="inner")
scorelist

region 1   region 2   region 3
RMSE            37.579422   0.893099  40.029709
SCORE            0.274239   0.999625   0.196614
PREDICTED MEAN  92.592568  68.728547  94.965046
TARGET MEAN     92.500000  68.825000  95.000000

### Вывод
- Второй регион показал наибольшие перспективы для разработки с точки зрения предсказания целевых значений, но тем не менее мы не можем рекомендовать его для разработки ввиду неоптимальных абсолютных параметров.

<a href='#top'>Back to top</a>

<div class="alert alert-block alert-info">
<b>Совет: </b> Можно было собрать все статистики (среднее, риск, интервал) в одной таблице, так было бы проще сравнивать.
</div>

<a id='section_5'></a>
# 5. Общий вывод

- Данные, представленные для прогноза, хорошо подготовлены: не имеют дубликатов и пропусков.
- В результате моделирования получили следующие результаты:
    - Регион 1: RMSE : 37.6  ; Score 0.27
    - Регион 2: RMSE : 0.89  ; Score 1.00
    - Регион 3: RMSE : 40.03 ; Score 0.20
- Исходя из проведенных расчетов, заключили, что минимальный объем сырья для безубыточной разработки составляет 111 тыс барелей. К сожалению, показатели всех представленных регионов ниже обозначенного порога.
- Второй регион показал наибольшие перспективы для разработки с точки зрения предсказания целевых значений, но тем не менее мы не можем рекомендовать его для разработки ввиду неоптимальных абсолютных параметров. Для завершения анализа необходимо больше данных по каждому региону, либо больше дополнительных признаков.

<a href='#top'>Back to top</a>

<div class="alert alert-block alert-success">
<b>Успех:</b> Всегда приятно видеть проект, который заканчивается выводом.
</div>

<a id='section_end'></a>
# Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован

<a href='#top'>Back to top</a>